### PR TITLE
add button layout support on viber channel

### PR DESCRIPTION
**Pull request by rasoro**
add support to manage viber keyboard button layout to modify the appearance of rapidpro quickreplies buttons, following some specifications of [this guide](https://developers.viber.com/docs/tools/keyboard-examples/).

the settings can be applied in the channel, and the `channel.config` will be that way:

```json
{
  "auth_token": "4e2927....",
  "callback_domain": "c175-206-42-34-254.ngrok.io",
  "button_layout": {
    "bg_color": "#f7bb3f",
    "text": "<font color=\"#ffffff\">*</font>",
    "text_size": "large"
  }
}
```
`button_layout` is the setting that will be applied to each button that will be displayed.

more details https://github.com/rapidpro/rapidpro/issues/1491

### CODE REVIEW

**Comment by rowanseymour** [comment]
The `rows` and `columns` values determine the dimensions of an individual button.. I'm not sure I follow how having a single pair of values hardcoded on the channel would be that useful.

**Comment by rasoro** [comment]
I hope it got good

<details>
  <summary>some examples</summary>

191d9561-46a5-4258-a46a-eef2efe9cf21

bac0173d-2abc-4222-9dd2-27e026dc7909

5c755e90-3e4b-4f9f-ad6f-ce8f385759c5

703951ab-02b9-459b-a924-9a0fc6ec1580

b223115f-0159-46ba-ad25-4d8db7230c6e


c412e8d4-4629-4f6e-ab76-a09c4e67b8ed

</details>

File: handlers/viber/keyboard.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/nyaruka/courier/utils"
+)
+
+// KeyboardButton is button on a keyboard, see https://developers.viber.com/docs/tools/keyboards/#buttons-parameters
+type KeyboardButton struct {
+	ActionType string `json:"ActionType"`
+	ActionBody string `json:"ActionBody"`
+	Text       string `json:"Text"`
+	TextSize   string `json:"TextSize"`
+	Columns    string `json:"Columns,omitempty"`
```
[the remark this answers is not in the document]
Comment (rasoro) [reply handlers/viber/keyboard.go:16]: unfortunately the viber does not allow floats on columns

File: handlers/viber/keyboard.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	// paddingRunes to help to calculate the size of button column width
+	paddingRunes = 2
+)
+
+// NewKeyboardFromReplies create a keyboard from the given quick replies
+func NewKeyboardFromReplies(replies []string, buttonConfig map[string]interface{}) *Keyboard {
+	rows := utils.StringsToRows(replies, maxRows, maxRowRunes, paddingRunes)
+	buttons := []KeyboardButton{}
+
+	for i := range rows {
```
[the remark this answers is not in the document]
Comment (rasoro) [reply handlers/viber/keyboard.go:41]: as viber doesn't support floats values on columns, so I can extend `utils.StringsToRows` if you think it's more suitable

[the remark this answers is not in the document]
Comment (rowanseymour) [reply handlers/viber/keyboard.go:41]: I'm looking at `utils.StringsToRows` right now tho and it's not obvious how you'd change it to work the way we need here so maybe the viber handler should just have it's own function for figuring out rows/cols. I think how it needs to work is you try to see if the first 6 quick replies will fit on a row, if not try the first 3, then 2 then 1. And repeat until all quick replies are assigned to a row.

[the remark this answers is not in the document]
Comment (rasoro) [reply handlers/viber/keyboard.go:41]: sorry for the delay, i sent a commit right now, where i followed your recommendation. If you see anything that could be improved, let me know.

**Comment by matmsa27** [comment]
Hi @rowanseymour , is everything ok with this PR? do you have any other suggestions to improve this PR? thank you so much

File: handlers/viber/keyboard.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	textStyle := strings.TrimSpace(fmt.Sprint(buttonConfig["text"]))
+	textSize := strings.TrimSpace(fmt.Sprint(buttonConfig["text_size"]))
+
+	if len(bgColor) == 7 {
+		b.BgColor = bgColor
+	}
+	if strings.Contains(textStyle, "*") {
+		b.Text = strings.Replace(textStyle, "*", b.Text, 1)
+	}
+	if map[string]bool{"small": true, "large": true}[textSize] {
```
Comment (rowanseymour) [inline handlers/viber/keyboard.go:72]: make this a global var rather than instantiating on the fly

File: handlers/viber/keyboard.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+func (b *KeyboardButton) ApplyConfig(buttonConfig map[string]interface{}) {
+	bgColor := strings.TrimSpace(fmt.Sprint(buttonConfig["bg_color"]))
+	textStyle := strings.TrimSpace(fmt.Sprint(buttonConfig["text"]))
+	textSize := strings.TrimSpace(fmt.Sprint(buttonConfig["text_size"]))
+
+	if len(bgColor) == 7 {
+		b.BgColor = bgColor
+	}
+	if strings.Contains(textStyle, "*") {
+		b.Text = strings.Replace(textStyle, "*", b.Text, 1)
```
Comment (rowanseymour) [inline handlers/viber/keyboard.go:70]: Does this mean we need to escape HTML chars here? I'm guessing this breaks for quick replies that contain `<` or `>`

File: handlers/viber/keyboard.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+	colsByRow := []int{6, 3, 2, 1}
+	i := 0
+
+	for len(strs) > 0 {
+		if len(strs) >= colsByRow[i] {
+			rowRunes = 0
+			for _, str := range strs[:colsByRow[i]] {
+				strRunes := utf8.RuneCountInString(str) + paddingRunes*2
+				rowRunes += strRunes
```
Comment (rowanseymour) [inline handlers/viber/keyboard.go:91]:
```
rowRunes += utf8.RuneCountInString(str) + paddingRunes*2
```

**Review by rowanseymour — CHANGES_REQUESTED** [review]

**Comment by rowanseymour** [comment]
@matmsa27 hey sorry just hit that re-review button after making changes and I'll know to take another look

File: handlers/viber/keyboard.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+const (
+	// maxColumns refer to the maximum number of columns units in a row
+	maxColumns = 6
+	// maxRowsRunes refer to maximum number of runes in the same row to help define buttons layout
+	maxRowRunes = 38
+	// paddingRunes to help to calculate the size of button column width
+	paddingRunes = 2
+)
+
+var textSizes = map[string]string{"small": "small", "regular": "regular", "large": "large"}
```
Comment (rowanseymour) [inline handlers/viber/keyboard.go:36]: if you're using a map to check something is in a set, use `map[string]bool`

**Review by rowanseymour — CHANGES_REQUESTED** [review]
one tweak and maybe you missed https://github.com/nyaruka/courier/pull/385/files#r759667007

File: handlers/viber/keyboard.go (hunk 4 above)
Comment (rasoro) [reply handlers/viber/keyboard.go:74]: I think `html.ScapeString` can prevent problems with HTML chars, so i put on line 51. It's fine, that way?
https://github.com/nyaruka/courier/pull/385/files#diff-f16d69b49e81106d40a4c7ff1068ddd7b399e1b38b29c48c0d3481897479679bR51

File: handlers/viber/keyboard.go (hunk 6 above)
Comment (rasoro) [reply handlers/viber/keyboard.go]: Sure, i got it. Thanks.

**Review by rowanseymour — APPROVED** [review]

**Pull request merged**